File: frontend/src/modules/chat/components/WorkflowPanel/documentPublicationUrl.test.ts
import { describe, expect, it } from 'vitest';
import { documentPublicationTargetUrl, documentPublicationUrl } from './documentPublicationUrl';

describe('document publication links', () => {
  it.each([
    ['feishu', { uri: 'feishu:/~doc/FixtureDoc' }, 'https://feishu.cn/docs/FixtureDoc'],
    ['notion', { doc_id: '12345678-1234-1234-1234-123456789abc' }, 'https://www.notion.so/12345678123412341234123456789abc'],
    ['feishu', { uri: 'feishu:/~docx/token?redirect=elsewhere' }, undefined],
    ['feishu', { doc_id: '../wrong-target' }, undefined],
    ['notion', { uri: 'notion:/~page/not-a-page-id' }, undefined],
    ['github', { uri: 'feishu:/~docx/FixtureDoc' }, undefined],
  ])('resolves only identified internal documents for %s', (provider, target, expected) => {
    expect(documentPublicationTargetUrl(target, provider)).toBe(expected);
  });
  it('does not present the WeChat console as the published draft', () => {
    expect(documentPublicationTargetUrl({ meta: { browser_url: 'https://mp.weixin.qq.com/' } }, 'wechat')).toBeUndefined();
  });
  it.each([
    'javascript:alert(1)', 'file:///tmp/note.md', 'https://user:password@example.test',
    'obsidian://vlt_fixture/note.md', 'obsidian://new?path=%2Ftmp%2Fnote.md',
    'obsidian://open?path=relative.md', 'obsidian://open?path=%2Ftmp%2Fnote.md&append=changed',
    'obsidian://open?path=%2Ftmp%2Fnote.md&x-success=https://example.test',
    'obsidian://open?path=%2Ftmp%2Fnote.md&path=%2Ftmp%2Fother.md',
    'obsidian://open?path=%2Ftmp%2Fnote%00.md',
  ])('does not expose an unusable or destructive target: %s', value => {
    expect(documentPublicationUrl(value)).toBeUndefined();
  });
  it('uses a valid browser URL when the PR URL is invalid', () => {
    expect(documentPublicationTargetUrl({ meta: { pull_request_url: 'javascript:alert(1)', browser_url: 'https://github.com/example/docs/pull/7' } }, 'github'))
      .toBe('https://github.com/example/docs/pull/7');
  });
  it('encodes Windows host paths and does not infer a note from an internal vault ID', () => {
    const value = documentPublicationTargetUrl({ meta: { local_path: 'C:\\My Vault\\中文 #1.md' } }, 'obsidian');
    expect(new URL(value!).searchParams.get('path')).toBe('C:\\My Vault\\中文 #1.md');
    expect(documentPublicationTargetUrl({ uri: 'obsidian://vlt_fixture/note.md' }, 'obsidian')).toBeUndefined();
    expect(documentPublicationTargetUrl({ meta: { local_path: 'relative.md' } }, 'obsidian')).toBeUndefined();
  });
});

File: backend/core/workflow/document_publication_url_test.go
package workflow

import (
	"encoding/json"
	"testing"
	"time"
)

func TestPublicationStatusTargetURL(t *testing.T) {
	for _, tt := range []struct{ name, provider, target, want string }{
		{"feishu", "feishu", `{"uri":"https://example.feishu.cn/docx/fixture"}`, "https://example.feishu.cn/docx/fixture"},
		{"feishu-wiki", "feishu", `{"uri":"https://example.feishu.cn/wiki/fixture"}`, "https://example.feishu.cn/wiki/fixture"},
		{"feishu-internal-docx", "feishu", `{"uri":"feishu:/~docx/FixtureDocumentToken","doc_id":"FixtureDocumentToken"}`, "https://feishu.cn/docx/FixtureDocumentToken"},
		{"feishu-internal-doc", "feishu", `{"uri":"feishu:/~doc/FixtureLegacyToken","doc_id":"FixtureLegacyToken"}`, "https://feishu.cn/docs/FixtureLegacyToken"},
		{"feishu-internal-wiki", "feishu", `{"uri":"feishu@FixtureSpace:/~node/FixtureWikiToken","doc_id":"FixtureDocumentToken"}`, "https://feishu.cn/wiki/FixtureWikiToken"},
		{"feishu-id-only", "feishu", `{"doc_id":"FixtureDocumentToken"}`, "https://feishu.cn/docx/FixtureDocumentToken"},
		{"notion-internal-page", "notion", `{"uri":"notion:/~page/12345678-1234-1234-1234-123456789abc","doc_id":"12345678-1234-1234-1234-123456789abc"}`, "https://www.notion.so/12345678123412341234123456789abc"},
		{"notion", "notion", `{"uri":"notion:/~page/fixture","meta":{"browser_url":"https://www.notion.so/fixture-document"}}`, "https://www.notion.so/fixture-document"},
		{"github-wiki", "github", `{"uri":"githubwiki://example/docs/Note","meta":{"browser_url":"https://github.com/example/docs/wiki/Note"}}`, "https://github.com/example/docs/wiki/Note"},
		{"github", "github", `{"uri":"github://repo/note.md","meta":{"browser_url":"https://github.com/example/docs/blob/main/note.md"}}`, "https://github.com/example/docs/blob/main/note.md"},
		{"github-pr", "github", `{"uri":"github://repo/note.md","meta":{"pull_request_url":"https://github.com/example/docs/pull/7","browser_url":"https://github.com/example/docs/blob/main/note.md"}}`, "https://github.com/example/docs/pull/7"},
		{"wechat", "wechat", `{"uri":"wechat://draft/fixture","meta":{"browser_url":"https://mp.weixin.qq.com/"}}`, ""},
		{"obsidian", "obsidian", `{"uri":"obsidian://vlt_fixture/note.md","meta":{"local_path":"/Users/test/My Vault/中文 #1.md"}}`, "obsidian://open?path=%2FUsers%2Ftest%2FMy%20Vault%2F%E4%B8%AD%E6%96%87%20%231.md"},
		{"obsidian-windows", "obsidian", `{"meta":{"local_path":"C:\\Notes\\中文.md"}}`, "obsidian://open?path=C%3A%5CNotes%5C%E4%B8%AD%E6%96%87.md"},
		{"obsidian-internal", "obsidian", `{"uri":"obsidian://vlt_fixture/note.md"}`, ""},
		{"unsafe", "github", `{"uri":"javascript:alert(1)"}`, ""},
		{"credentials", "github", `{"meta":{"browser_url":"https://user:password@github.com/example/docs"}}`, ""},
	} {
		t.Run(tt.name, func(t *testing.T) {
			for _, receipt := range []bool{false, true} {
				op := &DocumentPublicationOperation{Provider: tt.provider, TargetDocument: json.RawMessage(tt.target)}
				if receipt {
					op.TargetDocument = json.RawMessage(`{"uri":"https://example.test/stale"}`)
					op.ReceiptJSON, _ = json.Marshal(DocumentPublicationReceipt{TargetDocument: json.RawMessage(tt.target)})
				}
				if got := publicationStatus(op, time.Now()).TargetURL; got != tt.want {
					t.Errorf("receipt=%v: got %q, want %q", receipt, got, tt.want)
				}
			}
		})
	}
}
